lib/utility/init.ts

```diff
@@ -1,37 +1,56 @@
-import { MessagePortMain } from '@electron/internal/browser/message-port-main';
-import * as events from 'events';
+import { ParentPort } from '@electron/internal/utility/parent-port';
 const Module = require('module');
-const { EventEmitter } = events;
 const v8Util = process._linkedBinding('electron_common_v8_util');
 
+const entryScript: string = v8Util.getHiddenValue(process, '_serviceStartupScript');
 // We modified the original process.argv to let node.js load the init.js,
 // we need to restore it here.
-process.argv.splice(1, 1, process._serviceStartupScript);
+process.argv.splice(1, 1, entryScript);
 
 // Clear search paths.
 require('../common/reset-search-paths');
 
 // Import common settings.
 require('@electron/internal/common/init');
 
-const parentPort: Electron.ParentPort = new EventEmitter() as any;
-parentPort.postMessage = v8Util.getHiddenValue(process, '_postMessage');
+const parentPort: any = new ParentPort();
+Object.defineProperty(process, 'parentPort', {
+  enumerable: true,
+  writable: false,
+  value: parentPort
+});
 
-v8Util.setHiddenValue(global, 'messagechannel', {
-  didReceiveMessage (event: any, ports: any[]) {
-    event.ports = ports.map(p => new MessagePortMain(p));
-    (parentPort as events.EventEmitter).emit('message', event);
+// Based on third_party/electron_node/lib/internal/worker/io.js
+parentPort.on('newListener', (name: string) => {
+  if (name === 'message' && parentPort.listenerCount('message') === 0) {
+    parentPort.start();
   }
 });
 
-Object.defineProperty(process, 'parentPort', {
-  get () {
-    return parentPort;
-  },
-  enumerable: true
+parentPort.on('removeListener', (name: string) => {
+  if (name === 'message' && parentPort.listenerCount('message') === 0) {
+    parentPort.pause();
+  }
 });
 
+const origNewListener = parentPort[Symbol('kNewListener')];
+parentPort[Symbol('kNewListener')] = function (size: number, type: string, ...args: any[]) {
```

Comment:
Does this ... work?

```
> Symbol('kNewListener') === Symbol('kNewListener')
false
```

Reply:
This is an error in the implementation, I copied the overrides based on https://github.com/nodejs/node/blob/36805e852471cc7a24524a329848da5c8f163864/lib/internal/worker/io.js#L298-L313 but actually in our case `ParentPort` is just an `EventEmitter` and not an `EventTarget` whereas the Node.js implementation is one https://nodejs.org/dist/latest-v16.x/docs/api/events.html#eventtarget-and-event-api. I will remove the `EventTarget` overrides.

Reply:
Refs https://github.com/electron/electron/pull/34980/commits/a82d3223ede85df14b97780ef2bc1cce21e3aa39